### PR TITLE
[DT-038894] mongodb bulk step additions

### PR DESCRIPTION
**Pull request by karstendecisions**

### CODE REVIEW

File: Decisions.MongoDB/ReplaceBulkDocumentStep.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            var updates = ids
+                .Select(id => new ReplaceOneModel<TDocument>(
+                    FetchStepUtility.GetIdMatchFilter<TDocument>(id, idType),
+                    SetId<TDocument>(idPropertyInfo, documentInput, id)){ IsUpsert = Upsert })
+                .Cast<WriteModel<TDocument>>()
+                .ToList();
+
+            
+            var result = GetMongoCollection<TDocument>(data).BulkWriteAsync(
+                updates, new BulkWriteOptions() { IsOrdered = false }).Result;
```
Comment (karstendecisions) [inline Decisions.MongoDB/ReplaceBulkDocumentStep.cs]: still need to clean this part up (WIP)

File: Decisions.MongoDB/BaseDeleteStep.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+using DecisionsFramework.Design.ConfigurationStorage.Attributes;
+using DecisionsFramework.Design.Flow;
+using DecisionsFramework.Design.Flow.Mapping;
+
+namespace Decisions.MongoDB
+{
+    [Writable]
+    public abstract class BaseDeleteStep : BaseMongoDBStep, ISyncStep, IDataConsumer
+    { 
+
```
Comment (jason-decisions) [inline Decisions.MongoDB/BaseDeleteStep.cs]: Extra line break

File: Decisions.MongoDB/BaseMongoDBStep.cs [hunk 3]
```diff
@@ -307,6 +308,55 @@ internal static Type GetIdPropertyType(IdType idType)
...
+            
+            // this recursion by definition should never happen twice, we require array input but allow initial dyn call
+            if (!data.GetType().IsArray)
+                return GetInputData<TData>(new[] { data }, name);
+            
+            // the calling method should not pass the type representing an array but rather the target mapped element type
+            if (typeof(TData).IsArray)
+                throw new InvalidConstraintException($"Invalid type {typeof(TData).FullName}, element type expected");
+            
+
```
Comment (jason-decisions) [inline Decisions.MongoDB/BaseMongoDBStep.cs]: Extra line break

File: Decisions.MongoDB/BaseReplaceStep.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+using DecisionsFramework.Design.ConfigurationStorage.Attributes;
+using DecisionsFramework.Design.Flow;
+using DecisionsFramework.Design.Flow.Mapping;
+using DecisionsFramework.Design.Flow.Mapping.InputImpl;
+using DecisionsFramework.Design.Properties;
+
+namespace Decisions.MongoDB
+{
+    [Writable]
+    public abstract class BaseReplaceStep : BaseMongoDBStep, ISyncStep, IDataConsumer, IDefaultInputMappingStep
```
Comment (jason-decisions) [inline Decisions.MongoDB/BaseReplaceStep.cs]: Let's remove this class since we're removing the `ReplaceBulkDocumentStep` and go back to how it was before.

File: Decisions.MongoDB/DeleteBulkDocumentStep.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+using DecisionsFramework.Design.ConfigurationStorage.Attributes;
+using DecisionsFramework.Design.Flow;
+using DecisionsFramework.Design.Flow.Mapping;
+using MongoDB.Bson;
+using MongoDB.Driver;
+
+namespace Decisions.MongoDB
+{
+    [Writable]
+    public class DeleteBulkDocumentStep : BaseDeleteStep
```
Comment (jason-decisions) [inline Decisions.MongoDB/DeleteBulkDocumentStep.cs]: Let's rename this to `BulkDeleteDocumentStep`

File: Decisions.MongoDB/DeleteBulkDocumentStep.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            object inputIds = data[DOCUMENT_ID_INPUT_NAME];
+            if (inputIds == null || ((string[])inputIds).Length == 0)
+                throw new LoggedException($"{DOCUMENT_ID_INPUT_NAME} is missing");
+            List<string> inputs = new List<string>((string[])data[DOCUMENT_ID_INPUT_NAME]);
+            DeleteResult result = GetMongoRawDocumentCollection(data)
+                .DeleteMany(FetchStepUtility.GetIdsInFilter<BsonDocument>(inputs, GetIdPropertyTypeEnum()));
+            return result.DeletedCount == 0 ? new ResultData(PATH_ERROR) : new ResultData(PATH_SUCCESS);
+        }
+        
+
```
Comment (jason-decisions) [inline Decisions.MongoDB/DeleteBulkDocumentStep.cs]: Extra line breaks

File: Decisions.MongoDB/FetchStepUtility.cs [hunk 7]
```diff
@@ -144,6 +144,28 @@ internal static FilterDefinition<TDocument> GetIdMatchFilter<TDocument>(object i
             }
         }
 
+
+
```
Comment (jason-decisions) [inline Decisions.MongoDB/FetchStepUtility.cs]: Extra line breaks

File: Decisions.MongoDB/InsertBulkDocumentStep.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+using DecisionsFramework;
+using DecisionsFramework.Design.ConfigurationStorage.Attributes;
+using DecisionsFramework.Design.Flow;
+using DecisionsFramework.Design.Flow.Mapping;
+using MongoDB.Driver;
+
+namespace Decisions.MongoDB
+{
+    [Writable]
+    public class InsertBulkDocumentStep : BaseInsertStep
```
Comment (jason-decisions) [inline Decisions.MongoDB/InsertBulkDocumentStep.cs]: Let's rename this to `BulkInsertDocumentStep`

File: Decisions.MongoDB/DeleteBulkDocumentStep.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                inputs.Add(new DataDescription(GetIdPropertyType(), DOCUMENT_ID_INPUT_NAME, true));
+
+                return inputs.ToArray();
+            }
+        }
+
+        public override OutcomeScenarioData[] OutcomeScenarios => new[]
+        {
+            new OutcomeScenarioData(PATH_SUCCESS),
+            new OutcomeScenarioData(PATH_ERROR)
```
Comment (jason-decisions) [inline Decisions.MongoDB/DeleteBulkDocumentStep.cs]: Why is this the only step with an Error Path? We should probably just be consistent.

File: Decisions.MongoDB/ReplaceBulkDocumentStep.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+using System.Reflection;
+using DecisionsFramework;
+using DecisionsFramework.Design.Flow;
+using DecisionsFramework.Design.Flow.Mapping;
+using DecisionsFramework.Utilities;
+using MongoDB.Driver;
+
+namespace Decisions.MongoDB
+{
+    public class ReplaceBulkDocumentStep : BaseReplaceStep
```
Comment (jason-decisions) [inline Decisions.MongoDB/ReplaceBulkDocumentStep.cs]: Let's just get rid of this since there's no obvious method in Mongo's API to use.

File: Decisions.MongoDB/ReplaceDocumentStep.cs [hunk 11]
```diff
@@ -1,116 +1,66 @@
...
+using DecisionsFramework;
 
 namespace Decisions.MongoDB
 {
     [Writable]
-    public class ReplaceDocumentStep : BaseMongoDBStep, ISyncStep, IDataConsumer, IDefaultInputMappingStep
-    {
-        const string DOCUMENT_INPUT = "Document";
-        const string DOCUMENT_ID_INPUT = "Document ID";
+    public class ReplaceDocumentStep : BaseReplaceStep
```
Comment (jason-decisions) [inline Decisions.MongoDB/ReplaceDocumentStep.cs]: Let's undo these changes and just clean up the imports in this class.

File: Decisions.MongoDB/DeleteBulkDocumentStep.cs (hunk 5 above)
Comment (karstendecisions) [reply Decisions.MongoDB/DeleteBulkDocumentStep.cs]: I mean from an alphabetical order perspective, it makes sense to name them this way

File: Decisions.MongoDB/DeleteBulkDocumentStep.cs (hunk 9 above)
Comment (karstendecisions) [reply Decisions.MongoDB/DeleteBulkDocumentStep.cs]: No idea, the delete had it before so I copied in suit

**Review by jason-decisions — APPROVED** [review]

**Comment by karstendecisions** [comment]
Tested from fresh install upgrade 8.14 to manual new version.   An odd error appeared but it did not affect anything and went away after executing.  I would say functionality still works as expected
